### PR TITLE
Add simpleperf type to perf instrument

### PR DESCRIPTION
**Pull request by robF8**

### CODE REVIEW

**Comment by setrofim** [comment]
Please rebase onto the latest master.

**Comment by setrofim** [comment]
Please fix your commits -- you have multiple commits with the same title and no commit message. Please squash or split as necessary to make sure there is one logical change per commit, and the commit message describes what and why is being committed.

File: wa/instruments/perf.py [hunk 1]
```diff
@@ -31,51 +32,72 @@ class PerfInstrument(Instrument):
     name = 'perf'
     description = """
     Perf is a Linux profiling with performance counters.
+    Simpleperf is an Android profiling tool woth performance counters.
```
[the remark this answers is not in the document]
Comment (robF8) [reply wa/instruments/perf.py]: Good spot. Fixed in latest push

File: wa/instruments/perf.py [hunk 2]
```diff
@@ -134,5 +181,91 @@ def update_output(self, context):
...
+                                for i in range(1, len(column_headers)):
+                                    classifiers[column_headers[i]] = line[column_header_indeces[i]:column_header_indeces[i + 1]].strip()
+
+                                context.add_metric('{}_{}_Overhead'.format(label, event_type),
+                                                   numeric(words[0].strip('%')),
+                                                   'percent',
+                                                   classifiers=classifiers)
+
+    def _process_simpleperf_stat_output(self, context, outdir):
+        for opts, label in zip(self.option_string, self.labels):
```
[the remark this answers is not in the document]
Comment (robF8) [reply wa/instruments/perf.py]: Thanks, changed to search directory for labels. Labels are added if None on devlib side

**Comment by robF8** [comment]
Added updates based on feedback

File: wa/instruments/perf.py [hunk 3]
```diff
@@ -31,51 +32,73 @@ class PerfInstrument(Instrument):
...
+                  global_alias='perf_type', description="""Specifies which type of perf binaries
+                  to install. Use simpleperf for collecting perf data on android systems."""),
+        Parameter('command', kind=str, default='stat', allowed_values=['stat', 'record'],
+                  global_alias='perf_command', description="""Specifies which perf command to use. If in record mode
+                  report command will also be executed and results pulled from target along with raw data
+                  file"""),
+        Parameter('events', kind=list_of_strs, global_alias='perf_events',
                   description="""Specifies the events to be counted."""),
-        Parameter('optionstring', kind=list_or_string, default='-a',
+        Parameter('option_string', kind=list_or_string, default='-a',
```
Comment (setrofim) [inline wa/instruments/perf.py]: Renaming existing parameter will a compatibility breaking change, please undo.

File: wa/instruments/perf.py [hunk 4]
```diff
@@ -134,5 +182,94 @@ def update_output(self, context):
                                 metric = '{}_{}'.format(label, match.group(3))
                                 context.add_metric(metric, count, classifiers=classifiers)
 
-    def teardown(self, context):
-        self.collector.reset()
+    @staticmethod
+    def _process_perf_record_output(context, outdir):
```
Comment (setrofim) [inline wa/instruments/perf.py]: The methods below need to be refactored --  please reduce the level of nesting and make sure lines are of reasonable length (as a rule of thumb --  the entire line should be viewable in GitHub without scrolling.

File: wa/instruments/perf.py (hunk 3 above)
Comment (robF8) [reply wa/instruments/perf.py]: Ah sorry, reverted back to original.

File: wa/instruments/perf.py (hunk 4 above)
Comment (robF8) [reply wa/instruments/perf.py]: Refactored output processing related methods. Thanks, i'll make sure to adhere to this in future.

**Comment by robF8** [comment]
Updated based on further feedback. Refactored output processing related methods.

File: wa/instruments/perf.py [hunk 5]
```diff
@@ -123,16 +170,134 @@ def update_output(self, context):
...
-                                context.add_metric(metric, count, classifiers=classifiers)
+                            self._add_perf_stat_metric(line, label, context)
 
-    def teardown(self, context):
-        self.collector.reset()
+    @staticmethod
+    def _add_perf_stat_metric(line, label, context):
+        line = line.split('#')[0]  # comment
+        match = PERF_COUNT_REGEX.search(line)
+        if match:
```
Comment (setrofim) [inline wa/instruments/perf.py]: With checks such as this, you can get rid of an indent level by reversing the logic:

```
if not match:
     return   # or "continue" inside a loop
# the rest of the code does not need to be indented after "if".
```

**Comment by robF8** [comment]
Hold off on more reviews, I need to make another small change. After talking with UXPERF team, I need to add another classifier to the stat output

**Comment by robF8** [comment]
Updated - added scaled percentage classifier to results parser.

**Pull request merged**